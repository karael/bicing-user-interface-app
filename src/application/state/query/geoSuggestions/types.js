import {
  FAILURE, PENDING, START, SUCCESS,
} from 'application/state/stateConstants';
import { defineAction } from 'redux-define';

export const FETCH = defineAction(
  'FETCH',
  [START, PENDING, SUCCESS, FAILURE],
  defineAction('query/geoSuggestions'),
);

export default FETCH;
