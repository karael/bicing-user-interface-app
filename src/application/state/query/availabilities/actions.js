import { FETCH } from 'application/state/query/availabilities/types';

export const fetchStart = (
  itineraryStep,
  itineraryAt,
  periodStartAt,
  periodEndAt,
  interval,
  ...stationIds
) => ({
  error: false,
  meta: { isFetching: true, itineraryStep, itineraryAt },
  payload: {
    periodStartAt, periodEndAt, interval, stationIds,
  },
  type: FETCH.START,
});
export const fetchPending = itineraryStep => ({
  error: false,
  meta: { isFetching: true, itineraryStep },
  type: FETCH.PENDING,
});
export const fetchSuccess = (itineraryStep, data) => ({
  error: false,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = (itineraryStep, error) => ({
  error: true,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.FAILURE,
  payload: error,
});
