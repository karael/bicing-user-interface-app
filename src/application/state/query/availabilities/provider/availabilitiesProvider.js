import availabilitiesClientFactory
  from 'application/state/query/availabilities/provider/availabilitiesClientFactory';

const availabilitiesProvider = (
  itineraryAt,
  periodStart,
  periodEnd,
  interval,
  stationIds,
) => availabilitiesClientFactory(
  itineraryAt,
  periodStart,
  periodEnd,
  interval,
  stationIds,
).find();

export default availabilitiesProvider;
