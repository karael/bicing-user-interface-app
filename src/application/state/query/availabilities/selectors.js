/* eslint-disable */
const selector = state => state.query.availabilities.itinerarySteps;

const step = (itineraryStep, state) =>
  selector(state).find((step) => step.itineraryStep === itineraryStep);

const data = (itineraryStep, state) =>
  step(itineraryStep, state) === undefined ? undefined : step(itineraryStep, state).data;

export const availabilityByItineraryStepAndId = (itineraryStep, id, state) => {
  if (isErrorByItineraryStep(itineraryStep, state) === true
    || step(itineraryStep, state) === undefined
    || step(itineraryStep, state).data === undefined) {
    return undefined;
  }

  return step(itineraryStep, state).data.find(model => model.id === id);
};

export const availabilityPropertyByKeyAndItineraryStepAndId = (state, key, itineraryStep, id) =>
  availabilityByItineraryStepAndId(itineraryStep, id, state) === undefined ?
    undefined :
    availabilityByItineraryStepAndId(itineraryStep, id, state)[key];

//////////////////////////////
//// Interfaced selectors ////
//////////////////////////////
export const itinerarySteps = state => [...new Set(selector(state).map(
  step => step.itineraryStep))
];

export const isErrorByItineraryStep = itineraryStep => state =>
  step(itineraryStep, state) === undefined ? false : step(itineraryStep, state).error;

export const errorByItineraryStep = itineraryStep => state =>
  isErrorByItineraryStep(itineraryStep)(state) === true
  && data(itineraryStep, state) instanceof Error ?
    data(itineraryStep, state).message : undefined;

export const isFetchingByItineraryStep = itineraryStep => state =>
  step(itineraryStep, state) === undefined ? false : step(itineraryStep, state).isFetching;

export const periodStartAtByItineraryStep = itineraryStep => state =>
  step(itineraryStep,state) === undefined ?
    undefined :
    step(itineraryStep, state).periodStartAt;

export const periodEndAtByItineraryStep = itineraryStep => state =>
  step(itineraryStep, state) === undefined ?
    undefined :
    step(itineraryStep, state).periodEndAt;

export const intervalByItineraryStep = itineraryStep => state =>
  step(itineraryStep, state) === undefined ?
    undefined :
    step(itineraryStep, state).interval;

export const itineraryAtByItineraryStep = itineraryStep => state =>
  step(itineraryStep, state) === undefined ?
    undefined :
    step(itineraryStep, state).itineraryAt;

export const statusByItineraryStepAndStationId = (itineraryStep, stationId) => state =>
  availabilityPropertyByKeyAndItineraryStepAndId(
    state,
    'status',
    itineraryStep,
    stationId);
export const availableBikeNumberByItineraryStepAndStationId = (itineraryStep, stationId) => state =>
  availabilityPropertyByKeyAndItineraryStepAndId(
    state,
    'availableBikeNumber',
    itineraryStep,
    stationId);
export const availableSlotNumberByItineraryStepAndStationId = (itineraryStep, stationId) => state =>
  availabilityPropertyByKeyAndItineraryStepAndId(
    state,
    'availableSlotNumber',
    itineraryStep,
    stationId);
