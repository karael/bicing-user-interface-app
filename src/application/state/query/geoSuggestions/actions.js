import { FETCH } from 'application/state/query/geoSuggestions/types';

export const fetchStart = () => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = geoSuggestions => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: geoSuggestions,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
