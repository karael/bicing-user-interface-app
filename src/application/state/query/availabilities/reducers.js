import { FETCH } from 'application/state/query/availabilities/types';
import produce from 'immer';
import { createReducer } from 'reduxsauce';

export const initialStateByItineraryStep = itineraryStep => ({
  data: undefined,
  error: false,
  isFetching: false,
  itineraryStep,
});

export const INITIAL_STATE = {
  itinerarySteps: [],
};

const findIndexByItineraryStep = (itineraryStep, itinerarySteps) => itinerarySteps.findIndex(
  step => step.itineraryStep === itineraryStep,
);

export const fetchStart = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  const { isFetching, itineraryStep } = action.meta;

  const draftByItineraryStep = initialStateByItineraryStep(itineraryStep);

  draftByItineraryStep.data = undefined;
  draftByItineraryStep.isFetching = isFetching;
  draftByItineraryStep.error = action.error;

  const index = findIndexByItineraryStep(itineraryStep, draft.itinerarySteps);

  if (index !== -1) {
    draft.itinerarySteps[index] = draftByItineraryStep;
  } else {
    draft.itinerarySteps.push(draftByItineraryStep);
  }
});

export const fetchPending = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  const index = findIndexByItineraryStep(action.meta.itineraryStep, draft.itinerarySteps);

  if (index !== -1) {
    const draftByItineraryStep = draft.itinerarySteps[index];

    draftByItineraryStep.data = undefined;
    draftByItineraryStep.isFetching = action.meta.isFetching;
    draftByItineraryStep.error = action.error;

    draft.itinerarySteps[index] = draftByItineraryStep;
  }
});

export const fetchSuccess = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  const index = findIndexByItineraryStep(action.meta.itineraryStep, draft.itinerarySteps);

  if (index !== -1) {
    const draftByItineraryStep = draft.itinerarySteps[index];

    draftByItineraryStep.data = action.payload;
    draftByItineraryStep.isFetching = action.meta.isFetching;
    draftByItineraryStep.error = action.error;

    draft.itinerarySteps[index] = draftByItineraryStep;
  }
});

export const fetchFailure = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  const index = findIndexByItineraryStep(action.meta.itineraryStep, draft.itinerarySteps);

  if (index !== -1) {
    const draftByItineraryStep = draft.itinerarySteps[index];

    draftByItineraryStep.data = action.payload;
    draftByItineraryStep.error = action.error;
    draftByItineraryStep.isFetching = action.meta.isFetching;
  }
});

const HANDLERS = {
  [FETCH.START]: fetchStart,
  [FETCH.PENDING]: fetchPending,
  [FETCH.SUCCESS]: fetchSuccess,
  [FETCH.FAILURE]: fetchFailure,
};

export default createReducer(INITIAL_STATE, HANDLERS);
