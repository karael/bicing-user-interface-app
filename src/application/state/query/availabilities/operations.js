import {
  fetchFailure,
  fetchPending,
  fetchSuccess,
} from 'application/state/query/availabilities/actions';
import availabilitiesProvider
  from 'application/state/query/availabilities/provider/availabilitiesProvider';
import { FETCH } from 'application/state/query/availabilities/types';
import { availabilitiesType } from 'domain/types/availabilityType';
import Joi from 'joi';
import { call, put, takeLatest } from 'redux-saga/effects';

export function* fetch(action) {
  const {
    meta: { itineraryAt, itineraryStep },
    payload: {
      periodStartAt, periodEndAt, interval, stationIds,
    },
  } = action;

  yield put(fetchPending(itineraryStep));

  try {
    const availabilities = yield call(availabilitiesProvider,
      itineraryAt,
      periodStartAt,
      periodEndAt,
      interval,
      stationIds);

    Joi.assert(availabilities, availabilitiesType);

    yield put(fetchSuccess(itineraryStep, availabilities));
  } catch (exception) {
    yield put(fetchFailure(itineraryStep, exception));
  }
}

export default function* operation() {
  yield takeLatest(FETCH.START, fetch);
}
