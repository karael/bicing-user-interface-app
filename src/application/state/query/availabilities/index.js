import * as actions from 'application/state/query/availabilities/actions';
import operation from 'application/state/query/availabilities/operations';
import reducer from 'application/state/query/availabilities/reducers';
import * as selectors from 'application/state/query/availabilities/selectors';
import * as types from 'application/state/query/availabilities/types';

export {
  actions,
  operation,
  selectors,
  types,
};

export default reducer;
