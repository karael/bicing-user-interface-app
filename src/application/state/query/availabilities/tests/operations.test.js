import * as actions from 'application/state/query/availabilities/actions';
import operation, { fetch } from 'application/state/query/availabilities/operations';
import availabilitiesProvider
  from 'application/state/query/availabilities/provider/availabilitiesProvider';
import { FETCH } from 'application/state/query/availabilities/types';
import { expectSaga, testSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';
import { v4 as uuid } from 'uuid';

describe('application/state/query/availabilities/operations', () => {
  test('should wait for a fetch start event to fetch with operation()', () => {
    testSaga(operation)
      .next()
      .takeLatestEffect(FETCH.START, fetch);
  });

  test('should dispatch a fetchPending action with fetch() generator', () => {
    const itineraryStep = 5;
    const action = {
      error: false,
      meta: { isFetching: true, itineraryStep, itineraryAt: '2017-08-08 12:22:12' },
      payload: {
        periodStartAt: '2017-08-08 12:12:12',
        periodEndAt: '2017-08-08 12:12:12',
        interval: '5T',
        stationIds: [],
      },
      type: FETCH.START,
    };
    testSaga(fetch, action)
      .next()
      .put(actions.fetchPending(itineraryStep));
  });

  test('should list expected availabilities with fetch() generator', () => {
    const fakeLastAvailabilities = [
      {
        '@id': '/api/last-availabilities-by-station/4028fe46-d10e-47df-9a66-f1960475af15',
        '@type': 'lastAvailabilitiesByStationView',
        id: '4028fe46-d10e-47df-9a66-f1960475af15',
        statedAt: '2018-09-19T12:50:03+02:00',
        availableBikeNumber: 1,
        availableSlotNumber: 25,
        status: 'OPENED',
      },
    ];
    const itineraryStep = 2;
    const action = {
      error: false,
      meta: { isFetching: true, itineraryStep, itineraryAt: '2017-08-08 12:22:12' },
      payload: {
        periodStartAt: '2017-08-08 12:12:12',
        periodEndAt: '2017-10-08 12:32:12',
        interval: '5T',
        stationIds: [uuid(), uuid()],
      },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [
          matchers.call.fn(availabilitiesProvider),
          fakeLastAvailabilities,
        ],
      ])
      .put(actions.fetchSuccess(itineraryStep, fakeLastAvailabilities))
      .run();
  });

  test('should handle error when api call response does not contains expected schema type', () => {
    const fakeLastAvailabilitiesWithMissingRequiredProperties = [{ availableBikeNumber: 1 }];
    const itineraryStep = 2;
    const action = {
      error: false,
      meta: { isFetching: true, itineraryStep, itineraryAt: '2017-08-08 12:22:12' },
      payload: {
        periodStartAt: '2017-08-08 12:12:12',
        periodEndAt: '2017-10-08 12:32:12',
        interval: '5T',
        stationIds: [
          '499cf165-9d02-4bfb-b150-6ba18f5a63fb',
          '1ee10d1e-5618-4441-a182-062dcf5fdaf1',
        ],
      },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [
          matchers.call.fn(availabilitiesProvider),
          fakeLastAvailabilitiesWithMissingRequiredProperties,
        ],
      ])
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });

  test('should handle error when api call failed in fetch() generator', () => {
    const error = new Error('error_api_call');
    const itineraryStep = 0;
    const action = {
      error: false,
      meta: { isFetching: true, itineraryStep, itineraryAt: '2017-08-08 12:22:12' },
      payload: {
        periodStartAt: '2017-08-08 12:12:12',
        periodEndAt: '2017-10-08 12:32:12',
        interval: '5T',
        stationIds: [uuid(), uuid()],
      },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [
          matchers.call.fn(availabilitiesProvider),
          throwError(error),
        ],
      ])
      .put(actions.fetchFailure(itineraryStep, error))
      .run();
  });
});
