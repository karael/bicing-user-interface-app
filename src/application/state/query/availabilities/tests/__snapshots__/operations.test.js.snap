// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/query/availabilities/operations should handle error when api call response does not contains expected schema type 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "2017-08-08 12:22:12",
          "2017-08-08 12:12:12",
          "2017-10-08 12:32:12",
          "5T",
          Array [
            "499cf165-9d02-4bfb-b150-6ba18f5a63fb",
            "1ee10d1e-5618-4441-a182-062dcf5fdaf1",
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/availabilitiesProvider",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
            "itineraryStep": 2,
          },
          "type": "query/availabilities/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
            "itineraryStep": 2,
          },
          "payload": Object {
            "_object": Array [
              Object {
                "availableBikeNumber": 1,
              },
            ],
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": "id",
                  "label": "id",
                },
                "message": "\\"id\\" is required",
                "path": Array [
                  0,
                  "id",
                ],
                "type": "any.required",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/availabilities/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;
